## TechnoHacks EduTech Data Analytics Internship
### Task 3 - Removing Duplicate Values
#### - By Azeen Hodekar                                             
##### Date - 16th September 2023

In [1]:
#Importing the libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#Loading the data in the dataframe
df = pd.read_csv('employee_data.csv', encoding='latin1')

In [3]:
#Head function to get overview of data
df.head()

EEID        Full Name                 Job Title Department  \
0  E02387      Emily Davis                Sr. Manger         IT   
1  E04105    Theodore Dinh       Technical Architect         IT   
2  E02572     Luna Sanders                  Director    Finance   
3  E02832  Penelope Jordan  Computer Systems Manager         IT   
4  E01639        Austin Vo               Sr. Analyst    Finance   

            Business Unit  Gender  Ethnicity  Age   Hire Date Annual Salary  \
0  Research & Development  Female      Black   55    4/8/2016     $141,604    
1           Manufacturing    Male      Asian   59  11/29/1997      $99,975    
2     Speciality Products  Female  Caucasian   50  10/26/2006     $163,099    
3           Manufacturing  Female  Caucasian   26   9/27/2019      $84,913    
4           Manufacturing    Male      Asian   55  11/20/1995      $95,409    

  Bonus %        Country       City   Exit Date  
0    15%   United States    Seattle  10/16/2021  
1     0%           China  Chongqing         NaN  
2    20%   United States    Chicago         NaN  
3     7%   United States    Chicago         NaN  
4     0%   United States    Phoenix         NaN

In [4]:
df.tail()

EEID     Full Name       Job Title  Department        Business Unit  \
995  E03094  Wesley Young     Sr. Analyst   Marketing  Speciality Products   
996  E01909  Lillian Khan         Analyst     Finance  Speciality Products   
997  E04398   Oliver Yang        Director   Marketing  Speciality Products   
998  E02521   Lily Nguyen     Sr. Analyst     Finance  Speciality Products   
999  E03545   Sofia Cheng  Vice President  Accounting            Corporate   

     Gender  Ethnicity  Age  Hire Date Annual Salary Bonus %        Country  \
995    Male  Caucasian   33  9/18/2016      $98,427      0%   United States   
996  Female      Asian   44  5/31/2010      $47,387      0%           China   
997    Male      Asian   31  6/10/2019     $176,710     15%   United States   
998  Female      Asian   33  1/28/2012      $95,960      0%           China   
999  Female      Asian   63  7/26/2020     $216,195     31%   United States   

         City Exit Date  
995  Columbus       NaN  
996   Chengdu  1/8/2018  
997     Miami       NaN  
998   Chengdu       NaN  
999     Miami       NaN

In [5]:
#Renaming the columns for better access
df.columns = [column.lower().replace(' ','_') for column in df.columns]

In [6]:
df.head()

eeid        full_name                 job_title department  \
0  E02387      Emily Davis                Sr. Manger         IT   
1  E04105    Theodore Dinh       Technical Architect         IT   
2  E02572     Luna Sanders                  Director    Finance   
3  E02832  Penelope Jordan  Computer Systems Manager         IT   
4  E01639        Austin Vo               Sr. Analyst    Finance   

            business_unit  gender  ethnicity  age   hire_date annual_salary  \
0  Research & Development  Female      Black   55    4/8/2016     $141,604    
1           Manufacturing    Male      Asian   59  11/29/1997      $99,975    
2     Speciality Products  Female  Caucasian   50  10/26/2006     $163,099    
3           Manufacturing  Female  Caucasian   26   9/27/2019      $84,913    
4           Manufacturing    Male      Asian   55  11/20/1995      $95,409    

  bonus_%        country       city   exit_date  
0    15%   United States    Seattle  10/16/2021  
1     0%           China  Chongqing         NaN  
2    20%   United States    Chicago         NaN  
3     7%   United States    Chicago         NaN  
4     0%   United States    Phoenix         NaN

In [7]:
#Changing datatype to datetime
df['hire_date'] = pd.to_datetime(df['hire_date'])

#Changing datatype to datetime
df['exit_date'] = pd.to_datetime(df['exit_date'])

#Removing '$' and ',' from the salary column and converting it to int 
df['annual_salary'] = df['annual_salary'].astype(str)
df['annual_salary'] = df['annual_salary'].str.replace(',', '')
df['annual_salary'] = df['annual_salary'].str.replace('$', '')
df['annual_salary'] = df['annual_salary'].astype(int)

#removing '%' from bonus column and converting it to int
df['bonus_%'] = df['bonus_%'].astype(str)
df['bonus_%'] = df['bonus_%'].str.replace('%','')
df['bonus_%'] = df['bonus_%'].astype(int)


In [8]:
#Checking datatype once again
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   eeid           1000 non-null   object        
 1   full_name      1000 non-null   object        
 2   job_title      1000 non-null   object        
 3   department     1000 non-null   object        
 4   business_unit  1000 non-null   object        
 5   gender         1000 non-null   object        
 6   ethnicity      1000 non-null   object        
 7   age            1000 non-null   int64         
 8   hire_date      1000 non-null   datetime64[ns]
 9   annual_salary  1000 non-null   int32         
 10  bonus_%        1000 non-null   int32         
 11  country        1000 non-null   object        
 12  city           1000 non-null   object        
 13  exit_date      85 non-null     datetime64[ns]
dtypes: datetime64[ns](2), int32(2), int64(1), object(9)
memory usage: 101.7+ 

Let us calculate shape before dropping duplicated values

In [9]:
df.shape

(1000, 14)

In [10]:
#Calculating summary stats for numeric columns
df.describe()

age                   hire_date  annual_salary      bonus_%  \
count  1000.000000                        1000    1000.000000  1000.000000   
mean     44.382000  2012-04-07 02:54:14.400000  113217.365000     8.866000   
min      25.000000         1992-01-09 00:00:00   40063.000000     0.000000   
25%      35.000000         2007-02-14 00:00:00   71430.250000     0.000000   
50%      45.000000         2014-02-15 12:00:00   96557.000000     0.000000   
75%      54.000000         2018-06-22 00:00:00  150782.250000    15.000000   
max      65.000000         2021-12-26 00:00:00  258498.000000    40.000000   
std      11.246981                         NaN   53545.985644    11.785625   

                           exit_date  
count                             85  
mean   2016-11-02 18:04:14.117647104  
min              1994-12-18 00:00:00  
25%              2014-12-25 00:00:00  
50%              2019-05-23 00:00:00  
75%              2021-04-09 00:00:00  
max              2022-08-17 00:00:00  
std                              NaN

Now, while calculating duplicated values, we must have some domain knowledge regarding the dataset. In this case, the dataset contains employee information of all employees. We must know that while some job titles might be repeated, they cannot be considered duplicates as many people can have the same job. We have to identify a column which has unique value for each employee, which in this case is the employee id (eeid). So while checking duplicate values we can use the subset argument as eeid in the duplicated() and drop_duplicates() methods.

In [11]:
df.duplicated(subset=['eeid'], keep='first')

0      False
1      False
2      False
3      False
4      False
       ...  
995    False
996    False
997    False
998    False
999    False
Length: 1000, dtype: bool

Let us count the number of duplicate values

In [12]:
df.duplicated(subset=['eeid'], keep='first').sum()

89

Now we drop the duplicate values using drop_duplicates() method and using eeid as value for the subset argument

In [13]:
df = df.drop_duplicates(subset=['eeid'], keep='first')

Now let us check the shape of the data set once again

In [14]:
df.shape

(911, 14)

This tells us that duplicate values have been dropped as the number of rows changed from 1000 to 911 (meaning 89  duplicate values have been dropped)

### Conclusion: Duplicate values in the employee id column have been removed